caPutLogApp/caPutJsonLogShellCommands.cpp

```diff
@@ -76,6 +77,26 @@ extern "C"
         caPutJsonLogShow(args[0].ival);
     }
 
+    /* Metadata */
+    int caPutJsonLogAddMetadata(char *property, char *value){
+        CaPutJsonLogTask *logger = CaPutJsonLogTask::getInstance();
+        std::string property_str(property);
+        std::string value_str(value);
+        if (logger != NULL) return logger->addMetadata(property_str, value_str);
+        else return -1;
+    }
+    static const iocshArg caPutJsonLogMetadataArg0 = {"property", iocshArgString};
+    static const iocshArg caPutJsonLogMetadataArg1 = {"value", iocshArgString};
+    static const iocshArg *const caPutJsonLogMetadataArgs[] =
+    {
+        &caPutJsonLogMetadataArg0,
+        &caPutJsonLogMetadataArg1
+    };
+    static const iocshFuncDef caPutJsonLogMetadataDef = {"caPutJsonLogMetadata", 2, caPutJsonLogMetadataArgs};
```

Comment:
I note that the C function that you call is `caPutJsonLogAddMetadata`, but the function exposed to the EPICS shell is `caPutJsonLogMetadata`. It would be nice to have those harmonised.

Reply:
@simon-ess Actually I believe the original is correct. `caPutJsonLogAddMetadata()` is the function that the VxWorks shell would call, and its argument list matches the arg's defined in `caPutJsonLogMetadataArgs`. The `iocshFuncDef caPutJsonLogMetadataDef` struct defines the iocsh command name, the arg array size and the array of pointers to `iocshArg`s that define their names and types.

Reply:
I'm not sure I follow. Does vxWorks actually call the C function directly?

My comment is more that most of the other function have the same shell name as the C function name; to wit
```c
static const iocshFuncDef caPutLogInitDef = {"caPutLogInit", 2, caPutLogInitArgs};
static void caPutLogInitCall(const iocshArgBuf *args)
{
    caPutLogInit(args[0].sval, args[1].ival);
}
```
but in comparison
```c
    static const iocshFuncDef caPutJsonLogMetadataDef = {"caPutJsonLogMetadata", 2, caPutJsonLogMetadataArgs};
    static void caPutJsonLogMetadataCall(const iocshArgBuf *args)
    {
        caPutJsonLogAddMetadata(args[0].sval, args[1].sval);
    }
```

Reply:
@simon-ess done

Reply:
Oops my bad, @simon-ess was right so thanks to @lucmaga for fixing that.

VxWorks provides its own target shell which does call the C functions directly, that's why we always have a separate function that iocsh calls and provides the args as an `iocshArgBuf`. RTEMS users also had a similar CExp shell written by Till Straumann for SLAC which did the same thing. We added iocsh and the extra arg-processing stuff when we needed to add support other OSs, but we kept backwards compatibility for those earlier shells. Both VxWorks and RTEMS now include iocsh and most RTEMS users run that now. Many VxWorks users still use the target shell, but may pass command strings to iocsh as that can expand environment variables in command-line arguments.